### PR TITLE
[Fix] Match `foreach` and per-parameter rounding in `SignSGD` and `Tiger`

### LINKED ISSUE #519: `foreach` step paths do not match the single-tensor paths in `SGDW`, `SignSGD` and `Tiger`
## Describe the bug

Some `foreach` step paths do not match their single-tensor counterparts. In `SGDW` the two compute
different recurrences. In `SignSGD` and `Tiger` they are equal in exact arithmetic but round differently,
and a `sign()` update turns that into a full step.

`foreach` is not opt-in: `BaseOptimizer.can_use_foreach` returns `True` whenever a group has dense, real
gradients, so the `foreach` path is the default on CPU as well. Passing `foreach=False` is the only way to
get the other one. Note also that `SignSGD`'s docstring says `None` means "auto-detect based on device
(True for CUDA, False otherwise)", which does not match that implementation.

For comparison, `torch.optim.SGD` builds its foreach path from the same operations in the same order as
the loop path: `buf.mul_(momentum).add_(grad, alpha=1 - dampening)` at `optim/sgd.py:361` versus
`_foreach_mul_` then `_foreach_add_` at `:440-441`. I ran `SGD` (momentum, dampening, Nesterov), `Adam`,
`Adam(amsgrad)`, `AdamW`, `RMSprop(centered)` and `Adamax` over 36 configurations each in float32 and
float64 and got bit-identical results either way. That is not universal in PyTorch: `ASGD`, `NAdam` and
`RAdam` differ at around `1e-7` in float32, and `SGD` itself differs in float16. Those updates are smooth,
so a small difference stays small, which is what separates them from the two cases below.

### A. `SGDW`: the paths compute different recurrences

`torch.optim.SGD` uses `buf = grad` on the first step, then `buf = momentum * buf + (1 - dampening) * grad`.

| # | Line | Problem |
|---|---|---|
| 1 | `sgd.py:184` | `state['momentum_buffer'] = p.clone()` seeds a gradient accumulator with parameter values |
| 2 | `sgd.py:214` | `_foreach_lerp_(buf, grad, 1 - dampening)` decays the buffer by `dampening`; `sgd.py:235` decays by `momentum` |
| 3 | `sgd.py:214` | at `dampening=0` the lerp weight is `1.0`, so `buf = grad` and momentum is dropped |
| 4 | `sgd.py:216` | the Nesterov branch writes into `grads`, but the update at `sgd.py:218` reads `momentum_buffers`, so it does not change that step; it does modify `p.grad` in place |
| 5 | `sgd.py:214` | the recurrence is unconditional, so `momentum=0, dampening>0` introduces momentum that was not asked for |

Item 1 reaches the update in the per-parameter path when `momentum > 0`, and in the `foreach` path when
`dampening != 0`. Items 2 and 3 apply when `momentum > 0`. `AccSGD` at `sgd.py:72` uses the same
`p.clone()` init, but its buffer is an auxiliary parameter sequence, so it is fine there; `SGDW` steps
along its buffer as a gradient (`sgd.py:235`, `:240`, `:251`).

Parameter trajectories are unaffected for `foreach` with `momentum=0, dampening=0`, and for the
per-parameter path with `momentum=0`, though the cloned state is still allocated.

### B. `SignSGD` and `Tiger`: same recurrence, different rounding

`_step_foreach` uses `_foreach_lerp_` while `_step_per_param` uses `buf.mul_().add_()`. The residue is
about `1e-8` in float32 and `1e-17` in float64. When the buffer lands near zero the two land on opposite
sides of it, the `sign()` calls disagree, and the paths step in opposite directions, up to `2 * lr` apart.

It needs near-cancellation. Over 200 random-gradient trajectories at `momentum=0.9` in float32 I saw no
difference. The size depends on the momentum value; at `lr=0.1` I see `2 * lr` at `momentum=0.1` and `lr`
at `momentum=0.9`.

## To Reproduce

* OS : macOS
* PyTorch version : 2.9.1
* Python version : 3.12
* pytorch-optimizer version : `main` (833817c)
* reproducible codes :

```python
import torch
from pytorch_optimizer import SGDW, SignSGD, Tiger


def run(cls, foreach, kwargs, grads, p0=5.0, lr=0.1):
    p = torch.nn.Parameter(torch.tensor([p0]))
    extra = {} if cls is torch.optim.SGD else {'foreach': foreach}
    opt = cls([p], lr=lr, weight_decay=0.0, **kwargs, **extra)
    for g in grads:
        p.grad = torch.tensor([g])
        opt.step()
    return round(p.item(), 6)


print('A. SGDW, grads = 1.0, 2.0, 3.0 from p = 5.0')
for kwargs in ({'momentum': 0.9, 'dampening': 0.0},
               {'momentum': 0.9, 'dampening': 0.0, 'nesterov': True},
               {'momentum': 0.9, 'dampening': 0.2},
               {'momentum': 0.0, 'dampening': 0.5}):
    g = (1.0, 2.0, 3.0)
    print(f'  {kwargs} -> per_param={run(SGDW, False, kwargs, g)} '
          f'foreach={run(SGDW, True, kwargs, g)} '
          f'torch_sgd={run(torch.optim.SGD, None, kwargs, g)}')

print('B. SignSGD and Tiger, grads = 1.0, -0.1 from p = 2.0')
for cls, kwargs in ((SignSGD, {'momentum': 0.1}), (Tiger, {'beta': 0.1})):
    g = (1.0, -0.1)
    pp = run(cls, False, kwargs, g, p0=2.0)
    fe = run(cls, True, kwargs, g, p0=2.0)
    print(f'  {cls.__name__} -> per_param={pp} foreach={fe} equal={pp == fe}')
```

Output:

```
A. SGDW, grads = 1.0, 2.0, 3.0 from p = 5.0
  {'momentum': 0.9, 'dampening': 0.0} -> per_param=2.8295 foreach=4.4 torch_sgd=4.049
  {'momentum': 0.9, 'dampening': 0.0, 'nesterov': True} -> per_param=2.44655 foreach=4.4 torch_sgd=3.5441
  {'momentum': 0.9, 'dampening': 0.2} -> per_param=3.0197 foreach=4.3448 torch_sgd=4.185
  {'momentum': 0.0, 'dampening': 0.5} -> per_param=4.4 foreach=4.175 torch_sgd=4.4
B. SignSGD and Tiger, grads = 1.0, -0.1 from p = 2.0
  SignSGD -> per_param=2.0 foreach=1.8 equal=False
  Tiger -> per_param=2.0 foreach=1.8 equal=False
```

In A, rows 1 and 2 differ only in `nesterov` and the `foreach` result is the same in both, which is
item 4. Row 4 sets `momentum=0`, and `foreach` still differs from both the per-parameter path and
`torch.optim.SGD`, which is item 5.

In B the second gradient cancels the buffer: after `1.0` the buffer is `0.9`, and
`0.9 * 0.1 + (-0.1) * 0.9` is exactly `0`.

While checking this I ran the same `foreach=True` versus `foreach=False` comparison across the optimizers
that expose the flag, and a few others differ at their default settings, `AdaBelief` and `StableAdamW` by
more than one learning rate. I did not chase those down and they may have separate causes. Happy to open a
PR with targeted tests, or to write the others up separately, if either would help.

### PR DESCRIPTION
**Pull request by 11NOel11**
## Problem (Why?)

Close #519 

Partly addresses #519, the `SignSGD` / `Tiger` half.

`_step_foreach` uses `_foreach_lerp_` while `_step_per_param` uses `mul_().add_()`. Same result in exact
arithmetic, different rounding, and `sign()` turns that into a full step when the buffer lands near zero.

## Solution (What/How?)

- [x] use `lerp_` in `_step_per_param` so both paths round the same way

One line in each. This keeps `_foreach_lerp_` as a single kernel. Moving the `foreach` path to
`_foreach_mul_` + `_foreach_add_` instead would work equally well if you would rather match
`torch.optim.SGD`.

`SGDW` is left alone here, since fixing it changes results for anyone using it with momentum.

## Other changes (bug fixes, small refactors)

N/A

## Notes

Added `test_sign_based_foreach_parity` for both optimizers. It fails on `main` and passes with this change.

## Checklist

- [x] Make sure to run `just format` before commit
- [x] My code adheres to the style guidelines of this project (`just check` shows no errors)
- [x] Both new and existing unit tests pass successfully on my local environment by running `just test`
- [ ] I have made the necessary changes to the documentation

`ruff check` is clean; I did not run `pyright`. Suite: 2494 passed before, 2496 after, same two
pre-existing failures.

### CODE REVIEW

**Review by kozistr — APPROVED** [review]
looks good to me!

afair, `_foreach_lerp_` landed in Pytorch 2.0, and MPS support was added in 2.4 (but, we've already used here). so maybe, it's time to bump the minimum pytorch version from 1.10 to 2.0.

anyway, thanks!
